offchain/hello-world-api/src/Signing.purs

```diff
@@ -0,0 +1,57 @@
+module Signing
+  ( getPubKey
+  , hsmSignTx
+  , getCmd
+  ) where
+
+import Contract.Prelude
+
+import Contract.Address (ByteArray)
+import Contract.Prim.ByteArray (byteArrayToHex)
+import Contract.Transaction (Ed25519Signature(..), PublicKey(..), Transaction(Transaction), TransactionWitnessSet, Vkey(..), _vkeys)
+import Ctl.Internal.Serialization (toBytes)
+import Ctl.Internal.Serialization as Serialization
+import Data.Lens (set)
+import Data.String (trim)
+import Effect.Aff (makeAff)
+import Effect.Exception (throw)
+import Node.Buffer.Class (toString)
+import Node.ChildProcess (ExecResult, defaultExecOptions, execFile)
+import Node.Encoding (Encoding(UTF8))
+import Node.Process (lookupEnv)
+import Untagged.Union (asOneOf)
+
+type Signer = ByteArray -> Aff String
+
+getCmd :: String -> Aff String
+getCmd varName = liftEffect $ do
+  lookupEnv varName >>= case _ of
+    Just cmd -> pure cmd
+    Nothing -> throw $ "expected " <> varName <> " to be set"
+
+getPubKey :: Aff PublicKey
+getPubKey = do
+  (res :: ExecResult) <- execAff "signer" [ "getPubKey" ]
+  PublicKey <<< trim <$> (liftEffect $ toString UTF8 res.stdout)
+
+hsmSignTx :: PublicKey -> Transaction -> Aff TransactionWitnessSet
+hsmSignTx = signTx cmdSigner
+
+cmdSigner :: Signer
+cmdSigner a = do
+  let args = [ "sign", byteArrayToHex a ]
+  (res :: ExecResult) <- execAff "signer" args
+  liftEffect $ trim <$> toString UTF8 res.stdout
+
+execAff :: String -> Array String -> Aff ExecResult
```

Comment:
I don't understand the purpose of wrapping `execFile` into `Aff`. `execFile` is intended to be used synchronously, you wait for the result, but then it's placed in an asynchronous context.

Reply:
I still dont think we need the Aff wrapper, we do want `yubihsm` to succeed before we continue doing anything else. I would simply use `liftEffect` in combination with [execFileSync](https://pursuit.purescript.org/packages/purescript-node-child-process/9.0.0/docs/Node.ChildProcess#v:execFileSync).

Reply:
Fair enough, the code is a bit nicer this way :+1:

Reply:
Though I'm pretty sure the call back was waiting for yubihsm to finish anyway FWIW

Reply:
It was, but I like to keep things simple